### PR TITLE
TransactionState::addCollection refactoring

### PR DESCRIPTION
**Pull request by graetzer**
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/794/

### CODE REVIEW

File: arangod/Aql/Query.h [hunk 1]
```diff
@@ -324,6 +329,9 @@ class Query {
 
   /// @brief pointer to vocbase the query runs in
   TRI_vocbase_t& _vocbase;
+  
+  /// @brief transaction context to use for this query
+  std::shared_ptr<transaction::Context> _transactionContext;
```
Comment (vasiliy-arangodb) [inline arangod/Aql/Query.h]: a transaction::Context is already part of Methods '_trx' why is it necessary to create another instance

File: arangod/StorageEngine/TransactionState.cpp [hunk 2]
```diff
@@ -119,6 +119,7 @@ TransactionState::Cookie::ptr TransactionState::cookie(
 
 /// @brief add a collection to a transaction
 int TransactionState::addCollection(TRI_voc_cid_t cid,
+                                    std::string const& cname,
```
Comment (vasiliy-arangodb) [inline arangod/StorageEngine/TransactionState.cpp]: why is a name necessary if a cid is already provided?

File: arangod/StorageEngine/TransactionState.cpp [hunk 3]
```diff
@@ -324,6 +325,7 @@ bool TransactionState::isExclusiveTransactionOnSingleCollection() const {
 }
 
 int TransactionState::checkCollectionPermission(TRI_voc_cid_t cid,
+                                                std::string const& cname,
```
Comment (vasiliy-arangodb) [inline arangod/StorageEngine/TransactionState.cpp]: why is a name necessary if a cid is already provided?

File: arangod/StorageEngine/TransactionState.h [hunk 4]
```diff
@@ -142,8 +142,8 @@ class TransactionState {
                                     AccessMode::Type accessType);
 
   /// @brief add a collection to a transaction
-  int addCollection(TRI_voc_cid_t cid, AccessMode::Type accessType,
-                    int nestingLevel, bool force);
+  int addCollection(TRI_voc_cid_t cid, std::string const& cname,
```
Comment (vasiliy-arangodb) [inline arangod/StorageEngine/TransactionState.h]: why is a name necessary if a cid is already provided?

File: arangod/StorageEngine/TransactionState.h [hunk 5]
```diff
@@ -201,14 +202,20 @@ class TransactionState {
   /// @brief whether or not a transaction is an exclusive transaction on a single collection
   bool isExclusiveTransactionOnSingleCollection() const;
 
-  int checkCollectionPermission(TRI_voc_cid_t cid, AccessMode::Type) const;
+  /// @brief check if current user can access this collection
+  int checkCollectionPermission(TRI_voc_cid_t cid,
+                                std::string const& cname,
```
Comment (vasiliy-arangodb) [inline arangod/StorageEngine/TransactionState.h]: why is a name necessary if a cid is already provided?

File: arangod/StorageEngine/TransactionState.h [hunk 6]
```diff
@@ -224,18 +231,18 @@ class TransactionState {
 
   ServerState::RoleEnum const _serverRole;  // role of the server
 
-  CollectionNameResolver const& _resolver;
```
Comment (vasiliy-arangodb) [inline arangod/StorageEngine/TransactionState.h]: this might be required by users of TransactionState to resolve data-source names, e.g. IResearchVinewNode

File: arangod/Transaction/Methods.cpp [hunk 7]
```diff
@@ -633,17 +633,38 @@ transaction::Methods::Methods(
...
+    TRI_ASSERT(_state->isTopLevelTransaction());
+    
     // register the transaction in the context
     _transactionContextPtr->registerTransaction(_state);
   }
 
   TRI_ASSERT(_state != nullptr);
 }
 
+/// @brief create the transaction, used to be UserTransaction
```
Comment (vasiliy-arangodb) [inline arangod/Transaction/Methods.cpp]: why the renaming?

File: arangod/Transaction/SmartContext.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  void unregisterTransaction() noexcept override;
+
+  /// @brief whether or not the transaction is embeddable
+  bool isEmbeddable() const override;
+  
+  static std::shared_ptr<Context> Create(TRI_vocbase_t&);
+  
+private:
+  /// @brief managed TransactionState
+  TransactionState *_state;
```
Comment (vasiliy-arangodb) [inline arangod/Transaction/SmartContext.h]: why must a transaction::Context know about a specific transaction state
IIRC a transaction::Context can be used with 0..n different TransactionStates

File: arangod/Utils/SingleCollectionTransaction.h [hunk 9]
```diff
@@ -38,11 +38,18 @@ class SingleCollectionTransaction final : public transaction::Methods {
 
  public:
   //////////////////////////////////////////////////////////////////////////////
-  /// @brief create the transaction, using a collection id
+  /// @brief create the transaction, using a collection
   //////////////////////////////////////////////////////////////////////////////
 
   SingleCollectionTransaction(std::shared_ptr<transaction::Context> const&,
-                              TRI_voc_cid_t, AccessMode::Type);
+                              LogicalCollection const*, AccessMode::Type);
```
Comment (vasiliy-arangodb) [inline arangod/Utils/SingleCollectionTransaction.h]: it is not advisable to pass raw pointers if a value is actually expected to always be non-null
otherwise it requires use of extra null checks, a reference might be more beneficial

File: arangod/Utils/SingleCollectionTransaction.h [hunk 10]
```diff
@@ -38,11 +38,18 @@ class SingleCollectionTransaction final : public transaction::Methods {
...
+  
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief create the transaction, using a view
+  //////////////////////////////////////////////////////////////////////////////
+  
+  SingleCollectionTransaction(std::shared_ptr<transaction::Context> const&,
```
Comment (vasiliy-arangodb) [inline arangod/Utils/SingleCollectionTransaction.h]: why is this required instead of maybe using a DataSource const&, if at all?

File: arangod/StorageEngine/TransactionState.h [hunk 11]
```diff
@@ -160,6 +160,7 @@ class TransactionState {
   /// @brief release collection locks for a transaction
   int unuseCollections(int nestingLevel);
 
+  /// FIXME delete, server-based locking should take care of this
   int lockCollections();
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/StorageEngine/TransactionState.h]: I wamtee to figure it out with michael

File: arangod/Transaction/Methods.cpp [hunk 12]
```diff
@@ -3243,18 +3248,17 @@ Result transaction::Methods::addCollectionToplevel(TRI_voc_cid_t cid,
...
+    res = _state->addCollection(cid, cname, type, _state->nestingLevel(), false);
   }
 
   if (res != TRI_ERROR_NO_ERROR) {
     if (res == TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION) {
       // special error message to indicate which collection was undeclared
       THROW_ARANGO_EXCEPTION_MESSAGE(
-          res, std::string(TRI_errno_string(res)) + ": " +
-                   resolver()->getCollectionNameCluster(cid) + " [" +
+          res, std::string(TRI_errno_string(res)) + ": " + cname + " [" +
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/Transaction/Methods.cpp]: yes

File: arangod/StorageEngine/TransactionState.cpp (hunk 3 above)
Comment (graetzer) [reply arangod/StorageEngine/TransactionState.cpp]: to get rid of the name resolution

File: arangod/Transaction/Methods.cpp (hunk 7 above)
Comment (graetzer) [reply arangod/Transaction/Methods.cpp]: because the UserTransaction was just a constructor

File: arangod/Transaction/SmartContext.h (hunk 8 above)
Comment (graetzer) [reply arangod/Transaction/SmartContext.h]: that might be accurate for the StandaloneContext, but not for this one. This is essentially a shared context in so far as it will get you the same TransactionStates every time. This allows us to share the TransactionStates between AQL snippets and TraverserEngines without having to actually control who starts the Transaction first

File: arangod/Utils/SingleCollectionTransaction.h (hunk 9 above)
Comment (graetzer) [reply arangod/Utils/SingleCollectionTransaction.h]: agreed, but now that I changed about 90 different files I would like to avoid going back and doing it again

File: arangod/Aql/Query.h (hunk 1 above)
Comment (graetzer) [reply arangod/Aql/Query.h]: its optional so we can use the SmartContext

File: arangod/StorageEngine/TransactionState.cpp (hunk 2 above)
Comment (graetzer) [reply arangod/StorageEngine/TransactionState.cpp]: to get rid of the name resolution in TransactionState

File: arangod/StorageEngine/TransactionState.h (hunk 6 above)
Comment (graetzer) [reply arangod/StorageEngine/TransactionState.h]: everything else usually just ueses the transaction::Methods interface directly
